tdvt/tdvt/tdvt.py

```diff
@@ -425,51 +467,84 @@ def init():
     active = 0
     for t in threads:
         if t.is_alive():
             active += 1
     return active
 
+
+def test_runner(all_tests, test_queue, max_threads):
+    for i in range(0, max_threads):
+        worker = threading.Thread(target=do_test_queue_work, args=(i, test_queue))
+        worker.setDaemon(True)
+        worker.start()
+    test_queue.join()
+    failed_tests = 0
+    total_tests = 0
+    for work in all_tests:
+        if work.copy_files_and_cleanup():
+            print("Left temp dir: " + work.temp_dir)
+        failed_tests += work.failed_tests if work.failed_tests else 0
+        total_tests += work.total_tests if work.total_tests else 0
+    return failed_tests, total_tests
+
+
 def run_tests_impl(tests, max_threads, args):
+    smoke_test_queue = queue.Queue()
+    smoke_tests = []
     test_queue = queue.Queue()
     all_work = []
     lock = threading.Lock()
-    logging.debug("test queue size is " + str(len(tests)))
+
     for test_set, test_config in tests:
         runner = TestRunner(test_set, test_config, lock, args.verbose, len(all_work) + 1)
-        all_work.append(runner)
-        test_queue.put(runner)
+        if test_set.smoke_test:
+            smoke_tests.append(runner)
+            smoke_test_queue.put(runner)
+        else:
+            all_work.append(runner)
+            test_queue.put(runner)
+
+    logging.debug("smoke test queue size is: " + str(len(smoke_tests)))
+    logging.debug("test queue size is: " + str(len(all_work)))
+
+    if not smoke_tests:
+        logging.warning("""No smoke tests detected. Tests will attempt to run without first verifying the data source connection. This may result in tests failing because of connection, rather than plugin, issues.""")
+        if args.smoke_test:
+            sys.exit(1)
 
     if not all_work:
-        print ("No tests found. Check arguments.")
+        print("No tests found. Check arguments.")
         sys.exit()
 
-    print ("Creating " + str(max_threads) + " worker threads.")
-    start_time = time.time()
-    for i in range(0, max_threads):
-        worker = threading.Thread(target=do_test_queue_work, args=(i, test_queue))
-        worker.setDaemon(True)
-        worker.start()
+    if smoke_tests:
+        print("Starting smoke tests. Creating", str(len(smoke_tests)), "worker threads.")
 
-    test_queue.join()
+        failed_smoke_tests, total_smoke_tests = test_runner(smoke_tests, smoke_test_queue, len(smoke_tests))
 
-    failed_tests = 0
-    total_tests = 0
-    for work in all_work:
-        if work.copy_files_and_cleanup():
-            print ("Left temp dir: " + work.temp_dir)
-        failed_tests += work.failed_tests if work.failed_tests else 0
-        total_tests += work.total_tests if work.total_tests else 0
+        if failed_smoke_tests > 0:
+            print("Test run aborted due to smoke test errors. {} smoke test(s) failed. Please check logs for information.".format(failed_smoke_tests))
```

Comment:
This is good for a single test suite like a plugin developer would use, but I think it might cause problems with our internal use case of testing many datasources at once. Can the smoke test be tied to the other test sets for the same suite? that way if postgres fails the smoke test then we only skip the postgres tests and not the mysql ones.

Reply:
I think this should be addressed in fce7f9c.

Doing a `tdvt --run postgres,athena` with athena broken, the smoke tests failed and the regular tests went on to run.

Reply:
![Snag_45e2bf](https://user-images.githubusercontent.com/5643640/61014361-ffb59900-a33b-11e9-8c50-d5523ee42f5d.png)